@here/harp-mapview/lib/ClipPlanesEvaluator.ts

```diff
@@ -14,6 +14,104 @@ import { MapViewUtils } from "./Utils";
 
+namespace SphericalProj {
+    const tmpVectors: THREE.Vector3[] = [new THREE.Vector3(), new THREE.Vector3()];
+
+    /**
+     * Calculate the horizon distance from a point above a sphere's surface.
+     *
+     * @remarks
+     * Returns zero if point is below surface or only very slightly above surface of sphere.
+     * @see https://en.wikipedia.org/wiki/Horizon#Derivation
+     * @param d - Distance from point to center of sphere.
+     * @param r - Radius of sphere.
+     */
+    function getHorizonDistance(d: number, r: number): number {
+        // There may be situations when maximum elevation still remains below sea level
+        // (elevation < 0) or it is negligible (elevation ~ epsilon)
+        return d - r < epsilon ? 0 : Math.sqrt(d * d - r * r);
+    }
+
+    /**
+     * Calculate furthest visible distance from camera position projected on eye direction.
+     *
+     * @remarks
+     * Furthest visible distance is assumed to be distance from camera to horizon
+     * plus distance from elevated geometry to horizon(so that high objects behind horizon
+     * remain visible).
+     * @see https://en.wikipedia.org/wiki/Horizon#Objects_above_the_horizon
+     * @param d - Distance from camera to origin.
+     * @param minR - Min sphere radius.
+     * @param maxR - Max sphere radius.
+     * @param alpha - Angle between camera eye vector and tangent.
+     */
+    export function getFarDistanceFromElevatedHorizon(
+        camera: THREE.PerspectiveCamera,
+        d: number,
+        minR: number,
+        maxR: number
+    ): number {
+        //         , - ~ ~ ~ - ,
+        //     , '               ' ,        E
+        //   ,           .           ,    . ' far + elev
+        //  ,            .   maxR    , '   /
+        // ,             .     ,  '    ,  /
+        // ,             . O '         , / te
+        // ,             | .           ,/
+        //  ,            |   . minR   ,/
+        //   ,           |      .    ,
+        //     ,         |        , '_____ far
+        //       ' -_, _ | _ ,  ' / T (Tangent point = Horizon)
+        //     near      |  eye   /
+        //             d |  +   / t
+        //               | +  /
+        //               |+ /
+        //               C ---> up
+
+        // Compute angle between camera eye vector and tangent in camera's up direction.
+        const normalToTanAngle = Math.asin(minR / d);
+        const eyeToTanAngle = Math.abs(normalToTanAngle - SphericalProj.getCameraPitch(camera));
+
+        const t = getHorizonDistance(d, minR);
+
+        // Because we would like to see elevated geometry that may be visible beyond
+        // the tangent point on ground surface, we need to extend viewing distance along
+        // the tangent line by te (see graph above).
+        const te = getHorizonDistance(maxR, minR);
+
+        // Project CE vector(length t + te) onto eye vector to get far distance.
+        const far = Math.cos(eyeToTanAngle) * (t + te);
```

Comment:
The previous explanation was perhaps too long, but at least it described how this formula was derived. I think it would be good to keep it there (or at least part of it). Especially the fact that the camera direction is length 1.

Reply:
I've added more details to the diagram and added an explanation here.